libraries/AP_BattMonitor/AP_BattMonitor_SMBus_SUI.cpp

```diff
@@ -0,0 +1,228 @@
+        _state.current_amps = -current_ma * 0.001;
+        _state.last_time_micros = tnow;
+    }
+
+    read_full_charge_capacity();
+
+    read_remaining_capacity();
+
+    // timeout after 5 seconds
+    if ((tnow - _state.last_time_micros) > AP_BATTMONITOR_SMBUS_TIMEOUT_MICROS) {
+        _state.healthy = false;
+        // do not attempt to ready any more data from battery
+        return;
+    } else {
+        _state.healthy = true;
+    }
+
+    // read the manufacturers data for button press
+    uint8_t mf_data[6];
+    if (read_block(BATTMONITOR_SMBUS_MANUFACTURE_DATA, mf_data, sizeof(mf_data))) {
+        bool pressed = (mf_data[1] >> 3) & 0x01;
+
+        if (button_press_count >= BATTMONITOR_SMBUS_SUI_BUTTON_DEBOUNCE) {
+            // battery will power off
+            AP_Notify::flags.powering_off = true;
+        } else if (pressed) {
+            // battery will power off if the button is held
+            button_press_count++;
+
+        } else {
+            // button released, reset counters
+            button_press_count = 0;
+            AP_Notify::flags.powering_off = false;
+        }
+    }
+
+    read_temp();
+
+    read_serial_number();
+    read_remaining_capacity();
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_temp(void)
+{
+    uint16_t data;
+    if (read_word(BATTMONITOR_SMBUS_TEMP, data)) {
+        float temp = ((float)(data - 2731)) * 0.1f;
+        _state.temperature = temp;
+        _state.temperature_time = AP_HAL::millis();
+    }
+
+    return true;
+}
+
+bool AP_BattMonitor_SMBus_SUI::read_remaining_capacity()
+{
+    if (_params._pack_capacity > 0) {
+        uint16_t data;
+        if (read_word(BATTMONITOR_SMBUS_REMAINING_CAPACITY, data)) {
+            float consumed = _params._pack_capacity - data;
+            if (consumed > 0 && consumed < _params._pack_capacity) {
+                _state.consumed_mah = consumed;
+                return true;
+            }
+        }
+    }
+
+    return false;
+}
+
+
+// read_block - returns true if successful
+bool AP_BattMonitor_SMBus_SUI::read_block(uint8_t reg, uint8_t* data, uint8_t len) const
+{
+    // buffer to hold results (2 extra byte returned holding length and PEC)
+    uint8_t buff[len+2];
+
+    // read bytes
+    if (!_dev->read_registers(reg, buff, sizeof(buff))) {
+        return false;
+    }
+
+    // get length
+    uint8_t bufflen = buff[0];
+
+    // sanity check length returned by smbus
+    if (bufflen == 0 || bufflen > len) {
+        return false;
+    }
+
+    // check PEC
+    uint8_t pec = get_PEC(AP_BATTMONITOR_SMBUS_I2C_ADDR, reg, true, buff, bufflen+1);
+    if (pec != buff[bufflen+1]) {
+        return false;
+    }
+
+    // copy data (excluding PEC)
+    memcpy(data, &buff[1], bufflen);
+
+    // return success
+    return true;
+}
+
+// read_bare_block - returns true if successful
+bool AP_BattMonitor_SMBus_SUI::read_block_bare(uint8_t reg, uint8_t* data, uint8_t len) const
```

Comment:
Whoa, why is this needed? This looks to give up on all the PEC protection, which doesn't make sense...